hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/slowlog/SlowLogEventHandler.java

```diff
@@ -0,0 +1,137 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.regionserver.slowlog;
+
+import com.lmax.disruptor.EventHandler;
+import com.lmax.disruptor.RingBuffer;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.hbase.shaded.protobuf.generated.TooSlowLog.SlowLogPayload;
+
+/**
+ * Event Handler run by disruptor ringbuffer consumer
+ */
+@InterfaceAudience.Private
+class SlowLogEventHandler implements EventHandler<RingBufferTruck> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SlowLogEventHandler.class);
+
+  private final SlowLogPayload[] slowLogPayloads;
+
+  private AtomicInteger slowLogPayloadIdx = new AtomicInteger();
+
+  private static final Lock LOCK = new ReentrantLock();
+
+  private static final int WRITE_LOCK_WAIT = 1;
+  private static final int READ_LOCK_WAIT = 1;
+
+  SlowLogEventHandler(int eventCount) {
+    this.slowLogPayloads = new SlowLogPayload[eventCount];
+  }
+
+  /**
+   * Called when a publisher has published an event to the {@link RingBuffer}
+   *
+   * @param event published to the {@link RingBuffer}
+   * @param sequence of the event being processed
+   * @param endOfBatch flag to indicate if this is the last event in a batch from
+   *   the {@link RingBuffer}
+   * @throws Exception if the EventHandler would like the exception handled further up the chain
+   */
+  @Override
+  public void onEvent(RingBufferTruck event, long sequence, boolean endOfBatch) throws Exception {
+    SlowLogPayload slowLogPayload = event.getPayload();
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Received Slow Log Event. RingBuffer sequence: {}, isEndOfBatch: {}, " +
+          "Event Call: {}", sequence, endOfBatch,
+        slowLogPayload.getCallDetails());
+    }
+    try {
+      if (LOCK.tryLock(WRITE_LOCK_WAIT, TimeUnit.SECONDS)) {
+        int index = slowLogPayloadIdx.getAndSet(
+          (slowLogPayloadIdx.get() + 1) % slowLogPayloads.length);
+        slowLogPayloads[index] = slowLogPayload;
+      }
```

Comment:
Based on the nature of ring buffer, we anyways are going to lose some imp debug log which happened long time back hence we are not much concerned about logging every single detail here, even if we miss a couple of them, it should not be end of the world IMHO.
What we do want to log is complete info of ongoing RPC calls and most importantly identify one of those regular slow log events based on which we can make further decisions. With ring buffer, anyways we might lose one-off cases. Hence, as part of sub-task, we are going to provide an option to user for saving all slow logs with complete info in HDFS files.

cc: @apurtell

Reply:
Logging is better than nothing; that satisfies my review query. I think the idea of a best-effort recording of observed events is a strange approach for a database to take. Updating a metric would also be useful for providing durable, light-weight account of these activities.

As for logging to HDFS files, I'm concerned that will cause more overhead and harm than good. We can take it up on the sub-task, but I'm currently thinking a better place would be a new system table.

Reply:
I kind of agree with @ndimiduk  here. Its odd that we are missing events.

I'm not super familiar with disruptor's implementation of ring buffer (which is obviously super complex and optimized for certain ingest pattern).. but I think we can simplify it by adding a simple thread-safe wrapper over EvictingQueue...

The wrapper runs in a simple thread that consumes input from a shared-buffer to which RpcServer writes and this thread keeps polling entries and appends to an underlying EvictingQueue...that way 

- we don't miss any events and the whole thing is asynchronous too (since the RpcServer just writes the entry to the shared buffer).
- Most of the logic can be folded into the wrapper class (currently there is a lot of serialization stuff happening in the context of RpcServer).. this thread can also "log" entries.
- Fewer lines of code.

Thoughts?

Reply:
Also, looking at the implementation, we are always blocking on a lock, so all writes are pretty much serialized.. so I'd guess my suggestion won't be any less performant..(I think)..